google/cloud/internal/timer_queue.cc

```diff
@@ -46,80 +56,83 @@ void TimerQueue::Shutdown() {
-    auto predicate = [this, until, &lk] {
-      auto const ne = NextExpiration(lk);
-      return shutdown_ || ne < until ||
-             HasExpiredTimer(lk, std::chrono::system_clock::now());
+  // It is easier to understand this if you first ignore shutdown and empty
+  // timer queues.
+  //
+  // The threads play two different roles:
+  // - A single thread at a time plays the leader role.
+  // - All other threads are followers and block on cv_follower_.
+  //
+  // Once a timer expires the leader thread relinquishes its role and wakes up
+  // one follower thread to become the new leader. Only after doing so it
+  // runs the code to expire the timer.
+  //
+  // This is all complicated by shutdown. Basically all thread needs to wake
+  // up when the TQ is shutdown and one of them will expire all the timers.
+  //
+  std::unique_lock<std::mutex> lk(mu_);
+  auto is_leader = false;
+  while (!shutdown_) {
+    if (!is_leader && has_leader_) {
+      // The current thread becomes a follower until no leader is available.
+      cv_follower_.wait(lk, [this] { return shutdown_ || !has_leader_; });
+      continue;
+    }
+    // The is_leader flag allows us to restart this loop without worrying about
+    // electing new leaders.
+    is_leader = true;
+    has_leader_ = true;
+    if (timers_.empty()) {
+      cv_.wait(lk, [this] { return shutdown_ || !timers_.empty(); });
+      continue;
+    }
+    // Should a new timer appear that changes the "first" timer, we need to
+    // wake up and recompute the sleep time. But note that the leader thread
+    // does not need to relinquish that role to do so.
+    auto until = timers_.begin()->first;
+    auto predicate = [this, until] {
+      return shutdown_ || timers_.empty() || timers_.begin()->first != until;
     };
-    // TODO(#10512): Only wake one thread to service the next timer instead
-    // of all of them.
-    cv_.wait_until(lk, until, predicate);
-    if (shutdown_) return NotifyShutdown(std::move(lk));
-    auto now = std::chrono::system_clock::now();
-    if (HasExpiredTimer(lk, now)) {
-      ExpireTimer(std::move(lk), now);
+    if (cv_.wait_until(lk, until, predicate)) {
+      // Timers can be expired only if the CV exits due to a timeout.
+      continue;
     }
+    // if we get here we know that predicate() is false. Which implies that
+    // timers_ is not empty.
+    auto p = std::move(timers_.begin()->second);
+    timers_.erase(timers_.begin());
+    // Relinquish the leader role, release the mutex and then signal a follower.
+    has_leader_ = false;
+    is_leader = false;
+    lk.unlock();
+    // Elect a new leader (if available) to continue expiring timers.
+    cv_follower_.notify_one();
+    // This may run user code (in the continuations for the future). That may
+    // do all kinds of things, including calling back on this class to create
+    // new timers. We cannot hold the mutex while it is running.
+    p.set_value(std::chrono::system_clock::now());
```

Comment:
I still believe we should use `until` to satisfy the future.  It (1) bears some relationship to the expiring timer, rather than being some arbitrary point between scheduling and witnessing, and (2) removes the last call to `now()`.  It would also allow the tests to be more exacting and less flaky.

Reply:
Done.